warehouse/accounts/views.py

```diff
@@ -282,6 +283,10 @@ def two_factor_and_totp_validate(request, _form_class=TOTPAuthenticationForm):
                 .hexdigest()
                 .lower(),
             )
+
+            if not request.user.has_recovery_codes:
```

Comment:
We have the `two_factor_state` dict constructed above... I'm not sure if it costs us anything to call `request.user.has_recovery_codes`, but we might just replace this with `if not two_factor_state.get('has_recovery_codes', False):`